File: 2020/18/part1/main.cpp
#include <iostream>
#include <sstream>
#include <vector>

#define atoi(x) ( x & 0xf )

std::vector<char> parse_tokens(const std::string& input);
long compute(const std::vector<char>& tokens, size_t& pos);

int main() {
    std::string line;
    std::string token;
    long total = 0;
    while (std::getline(std::cin, line)) {
        std::stringstream in(line);
        line.clear();
        while (in >> token) {
            line += token;
        }
        auto tokens = parse_tokens(line);
        size_t pos = 0;
        total += compute(tokens, pos);
    }
    std::cout << total << std::endl;
}

std::vector<char> parse_tokens(const std::string& input) {
    std::vector<char> tokens;
    std::string _input = input;

    while (!_input.empty()) {
        tokens.push_back(_input[0]);
        _input = _input.substr(1);
    }

    return tokens;
}

long compute(const std::vector<char>& tokens, size_t& pos) {
    if (pos == (tokens.size() - 1)) {
        return atoi(tokens.at(pos));
    }

    long val;
    if (tokens.at(pos) == '(') {
        pos += 1;
        val = compute(tokens, pos);
    } else {
        val = atoi(tokens.at(pos));
    }

    size_t i = pos + 1;
    while (i < tokens.size()) {
        if (tokens.at(i) == ')') {
            pos = i;
            return val;
        }

        if (tokens.at(i) == '+') {
            i++;
            if (tokens.at(i) == '(') {
                i++;
                val += compute(tokens, i);
            } else {
                val += atoi(tokens.at(i));
            }
        } else {
            i++;
            if (tokens.at(i) == '(') {
                i++;
                val *= compute(tokens, i);
            } else {
                val *= atoi(tokens.at(i));
            }
        }
        i++;
    }
    return val;
}

File: 2020/18/part1/main.py
import sys


def compute(tokens, pos):
    if pos == len(tokens)-1:
        return int(tokens[pos])

    val = 0
    if tokens[pos] == '(':
        val, pos = compute(tokens, pos+1)
    else:
        val = int(tokens[pos])

    i = pos + 1
    while i < len(tokens):
        if tokens[i] == ')':
            return val, i

        if tokens[i] == '+':
            i += 1
            if tokens[i] == '(':
                v, i = compute(tokens, i+1)
                val += v
            else:
                val += int(tokens[i])
        else:
            i += 1
            if tokens[i] == '(':
                v, i = compute(tokens, i+1)
                val *= v
            else:
                val *= int(tokens[i])
        i += 1
    return val, -1


total = 0
for line in sys.stdin.readlines():
    tokens = ''.join(line.split(' ')).strip()
    val, _ = compute(tokens, 0)
    total += val
print(total)
